manifest/manifest.go

```diff
@@ -0,0 +1,161 @@
+package manifest
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/filecoin-project/go-f3/certs"
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/ipfs/go-datastore"
+	"golang.org/x/xerrors"
+)
+
+var (
+	// Default configuration for the EC Backend
+	DefaultEcConfig = &EcConfig{
+		ECFinality:       900,
+		CommiteeLookback: 5,
+		ECDelay:          30 * time.Second,
+
+		ECPeriod: 30 * time.Second,
+	}
+
+	DefaultGpbftConfig = &GpbftConfig{
+		Delta:                3 * time.Second,
+		DeltaBackOffExponent: 2.0,
+		MaxLookaheadRounds:   5,
+	}
+
+	DefaultGpbftOptions = []gpbft.Option{
+		gpbft.WithMaxLookaheadRounds(DefaultGpbftConfig.MaxLookaheadRounds),
+		gpbft.WithDelta(time.Duration(DefaultGpbftConfig.Delta) * time.Second),
+		gpbft.WithDeltaBackOffExponent(DefaultGpbftConfig.DeltaBackOffExponent),
+	}
+)
+
+type OnManifestChange func(ctx context.Context, prevManifest Manifest, errCh chan error)
+
+type ManifestProvider interface {
+	// Run starts any background tasks required for the operation
+	// of the manifest provider.
+	Run(context.Context, chan error)
+	// Returns the list of gpbft options to be used for gpbft configuration
+	GpbftOptions() []gpbft.Option
+	// Set callback to trigger to apply new manifests from F3.
+	SetManifestChangeCallback(OnManifestChange)
+	// Manifest accessor
+	Manifest() Manifest
+}
+
+type Version string
+
+type GpbftConfig struct {
+	Delta                time.Duration
+	DeltaBackOffExponent float64
+	MaxLookaheadRounds   uint64
+}
+
+type EcConfig struct {
+	ECFinality       int64
+	ECDelay          time.Duration
+	CommiteeLookback uint64
+
+	//Temporary
+	ECPeriod            time.Duration
+	ECBoostrapTimestamp time.Time
+}
+
+// Manifest identifies the specific configuration for
+// the F3 instance currently running.
+type Manifest struct {
+	// Sequence number of the manifest.
+	// This is used to identify if a new config needs to be applied
+	Sequence uint64
+	// Initial instance to use for gpbft
+	InitialInstance uint64
```

Comment:
This is the property of the whole F3, not only gpbft. We would have teardown F3 and certstore to change it.